x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java

```diff
@@ -53,11 +77,78 @@ public class IlmHealthIndicatorService implements HealthIndicatorService {
     );
 
     public static final String AUTOMATION_DISABLED_IMPACT_ID = "automation_disabled";
+    public static final String INDEX_STUCK_IMPACT_ID = "index_stuck";
+    public static final List<HealthIndicatorImpact> AUTOMATION_DISABLED_IMPACT = List.of(
+        new HealthIndicatorImpact(
+            NAME,
+            AUTOMATION_DISABLED_IMPACT_ID,
+            3,
+            "Automatic index lifecycle and data retention management is disabled. The performance and stability of the cluster "
+                + "could be impacted.",
+            List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)
+        )
+    );
+
+    public static final List<HealthIndicatorImpact> INDEX_STUCK_IMPACT = List.of(
+        new HealthIndicatorImpact(
+            NAME,
+            INDEX_STUCK_IMPACT_ID,
+            3,
+            "Some indices have been longer than expected on the same Index Lifecycle Management action.",
+            List.of(ImpactArea.INGEST, ImpactArea.SEARCH)
+        )
+    );
+
+    // TODO fix this value
+    public static final TimeValue ONE_DAY = TimeValue.timeValueDays(1);
+
+    static final Map<String, RuleConfig> RULES_BY_ACTION_CONFIG = Stream.of(
+        new RuleConfig(RolloverAction.NAME, ONE_DAY, 100, List.of(WaitForActiveShardsStep.NAME)),
+        new RuleConfig(MigrateAction.NAME, ONE_DAY, 100),
+        new RuleConfig(
+            SearchableSnapshotAction.NAME,
+            ONE_DAY,
+            100,
+            List.of(WaitForDataTierStep.NAME, WaitForIndexColorStep.NAME, WaitForNoFollowersStep.NAME)
+        ),
+        new RuleConfig(DeleteStep.NAME, ONE_DAY, 100, List.of(WaitForNoFollowersStep.NAME)),
+        new RuleConfig(ShrinkAction.NAME, ONE_DAY, 100, List.of(WaitForNoFollowersStep.NAME)),
+        new RuleConfig(AllocateAction.NAME, ONE_DAY, 100),
+        new RuleConfig(ForceMergeAction.NAME, ONE_DAY, 100, List.of(WaitForIndexColorStep.NAME)),
+        new RuleConfig(DownsampleAction.NAME, ONE_DAY, 100, List.of(WaitForNoFollowersStep.NAME))
+    ).collect(toMap(RuleConfig::action, Function.identity()));
+
+    public static final StuckIndicesRuleEvaluator ILM_RULE_EVALUATOR = new StuckIndicesRuleEvaluator(
+        RULES_BY_ACTION_CONFIG.values().stream().map(RuleConfig::toPredicate).toList()
+    );
+
+    static final Map<String, Function<String, Diagnosis.Definition>> ACTION_STUCK_DEFINITIONS = RULES_BY_ACTION_CONFIG.entrySet()
+        .stream()
+        .collect(
+            Collectors.toUnmodifiableMap(
+                Map.Entry::getKey,
+                entry -> ilmPolicyName -> new Diagnosis.Definition(
+                    NAME,
+                    "stuck_action:" + entry.getKey(),
+                    "Some indices managed by the policy ["
```

Comment:
I think you're right here. This will create a lot of diagnosis definitions... on the other hand, how can we notify the user about the affected resources? For instance, right now we have the resource `INDEX` and we could create a new resource `ILM_POLICY`, but it's impossible to create a link between them ("the indices [yyy, zzz, ...] are managed by the policy [XX]"), unless we create a new type of resource (`INDICES_MANAGED_BY_ILM`), but not sure if that's correct...